module/zfs/dmu_objset.c

```diff
@@ -1789,6 +1789,8 @@ dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
 
 	for (int i = 0;
 	    i < multilist_get_num_sublists(os->os_dirty_dnodes[txgoff]); i++) {
+		if (multilist_sublist_is_empty(os->os_dirty_dnodes[txgoff], i))
+			continue;
```

Comment:
Skipping empty sublists makes a ton of sense to me, but I'm concerned about doing it locklessly.  I completely agree that after we drop the lock, we can no trust the returned result.  However, without the lock there's no longer any kind of memory barrier to ensure coherency of a stable multilist across cpus either.  While possibly impossible in practice due to the timing, it seems like this could result dnodes being skipped if the sublist was incorrectly determined to be empty.  How much do we really gain by not taking these locks?

Reply:
Probably not dramatic, since it is once per txg per dataset per sublist.  I was hoping to rely on other synchronization points, like txg bump, which obviously should be locked already.  But if there is opinion it is not enough, I may return locking back.

Reply:
Unless we can quantify a meaningful performance win by removing the locks I'd personally prefer we keep them.  Just to be on the safe side.  Though let's see if we can get @prakashsurya's and @ahrens throughts on this.